clients/pkg/logentry/stages/stage.go

```diff
@@ -94,9 +94,16 @@ func (s stageProcessor) Run(in chan Entry) chan Entry {
 }
 
 func toStage(p Processor) Stage {
+	output := os.Stderr
+
+	if InspectWithId {
+		output = os.Stdout
+	}
```

Comment:
Not a fan of this approach; we should pick one and stick with it.
This should technically all be `stderr` I feel. Can you expand on your reasoning for this (the PR description didn't explain it sufficiently, I feel).

Reply:
If the output is interleaving with `stdout` I'd argue that the better option is to redirect `stdout` to `/dev/null` to make the analysis easier. I want to try avoid putting too much in the tool that folks could otherwise fix with standard UNIX text processing.

Reply:
I had the issue that stderr and stdout are interleaved, so somehow my console did not finish the full write. The problem arised from having the log lines in stdout and the inspect content in stderr. But inspect does not print the full line at any time, so again, not possible to find what was actually the offending entry.

I have two suggestions:
- Inspect also prints the full line, so if you run with --dry-run, you get every line twice, but once to stderr and once to stdout
- Inspect does print directly to stdout

Reply:
> somehow my console did not finish the full write

Can you unpack this a bit for me here? The `stdout` and `stderr` streams are buffered, so do you suspect there's a buffering issue?

In any case, an example would be quite illustrative I think - if you can.

Reply:
Ok, I can't reproduce the issue anymore, I will just remove that part for now, and get back to it, in case it reappears.

Reply:
Got it to reproduce, here's a screenshot: (pretty hard to copy paste colors)

<img width="1290" alt="Screenshot 2021-09-08 at 09 21 29" src="https://user-images.githubusercontent.com/554474/132464520-7a4e4d9d-c545-4d66-80f8-d5495572d884.png">

I believe that's because the logger writes the output in multiple steps:

```
                 fmt.Fprint(l.Writer, blue.Sprint(e.Timestamp.Format("2006-01-02T15:04:05.999999999-0700")))
		fmt.Fprint(l.Writer, "\t")
		fmt.Fprint(l.Writer, e.Id)
		fmt.Fprint(l.Writer, "\t")
		fmt.Fprint(l.Writer, yellow.Sprint(e.Labels.String()))
		fmt.Fprint(l.Writer, "\t")
		fmt.Fprint(l.Writer, e.Line)
		fmt.Fprint(l.Writer, "\n")
```

Reply:
I believe you left out the most important line from the code you copied:
`l.Flush()`

It'll only flush that content to the output buffer when `Flush()` is called, I believe.

Out of interest, why do you need `-dry-run` and `-inspect` at the same time?

Reply:
Because I need to validate the produced loglines, this whole thing ties into #4139 
I broke my loki instance already multiple times because either the logs produced had too many labels due to a regex mishap, so now I like to precheck what's going on ;)

Also, I have noticed #4288 during testing here.

Reply:
OK, confirmed now, the Flush or multiple writes has no impact on the interleaving.

If stderr and stdout is used at the same time, I got the interleaving.

Even with single line print, like so, the issue still happened:

```
fmt.Fprint(l.Writer, blue.Sprint(e.Timestamp.Format("2006-01-02T15:04:05.999999999-0700")) + "\t" + e.Id + " \t" + yellow.Sprint(e.Labels.String()) + "\t" + e.Line + "\n")
		l.Flush()
```

Reply:
Yeah, they will interleave because they are separate file descriptors with their own buffers, which both write the your terminal concurrently.

I think you should redirect your stdout and stderr to separate files for later analysis.
`<cmd> 1>/tmp/out.log 2>/tmp/err.log`

I'm not convinced we need any special handling for this in Promtail, although I do think we should make `dry-run` and `inspect` both output to `stderr`.

Reply:
Yes, that's why I changed it to the same descriptor.

And I think dry-run and inspect work together, using it without dry-run seems ... wrong?

So, I can change the writer for the Logger to stderr?

Reply:
Hhmm.
OK. Let's try that and see how it feels.

Pity I didn't notice this discrepancy when writing `-inspect` initially; sucks to be changing the interface but maybe this is our only option.